### ISSUE TITLE
Feature: Make macros public

### ISSUE THREAD

**Issue by MRuecklCC** [tokio-rs/axum#1431]
## Feature Request
Make marcos in toplevel marco.rs public to other crates.

### Motivation
I was just playing around with the custom extractors and rejections, and following along how this is done for JSON bodies in axum. I want to achieve something similar for other content types, and thought I can simply define my own rejections and composite rejections with the same macros. Unfortunately they are declared as `pub(crate)` so I cant get them :-(

I think they would be useful for other users as well. I assume pulling them into the public API would require some more documentation, besides that the change would be trivial(?).

### Alternatives
I understand, that having a larger public API comes with responsibilities on stability etc. So this should be a well agreed upon decision (in either direction).

**Comment by jplatte** [comment]
I don't think that's going to happen. I can't find an existing issue about this, but I'm pretty sure David has previously rejected a similar request.

The one thing in this direction I think we could add are `IntoResponseParts` and `IntoResponse` derive macros.

**Comment by davidpdrsn** [comment]
> I don't think that's going to happen. I can't find an existing issue about this, but I'm pretty sure David has previously rejected a similar request.

Yeah I don't think those macros should be public. Writing good `macro_rules!` macros is really hard so honestly I'd just recommend you copy/paste them into your code as necessary. I also don't think they're that necessary if you're just defining a couple of rejection types. But if you're defining dozens, like axum does, then they are indeed helpful.

**Comment by MRuecklCC** [comment]
Good :) Totally understand! Thanks for the quick response and keep up the awesome work! <3

**Comment by davidpdrsn** [comment]
> The one thing in this direction I think we could add are IntoResponseParts and IntoResponse derive macros.

I'm up for having something like that!

**Comment by MRuecklCC** [comment]
> The one thing in this direction I think we could add are IntoResponseParts and IntoResponse derive macros.

How would that look like? Following [the cookbook](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/procedural-macros.html) maybe like this?

```rust
#[derive(IntoResponse)]
#[status = UNSUPPORTED_MEDIA_TYPE]
#[body = "Expected request with `Content-Type: text/sql`"]
pub struct MyRejectionReason;
```

I would like that even more than the currently available macros :-)

**Comment by MRuecklCC** [comment]
And, yes, I am playing around with SQL in the body. :shrug:

**Comment by davidpdrsn** [comment]
> How would that look like?

No idea. Haven't thought about it.

> And, yes, I am playing around with SQL in the body. 

😱

**Comment by jplatte** [comment]
I was thinking about the composite rejection case:

```rust
#[derive(IntoResponse)]
pub enum JsonRejection {
    JsonDataError(JsonDataError),
    JsonSyntaxError(JsonSyntaxError),
    MissingJsonContentType(MissingJsonContentType),
    BytesRejection(BytesRejection),
}

// generates
impl IntoResponse for JsonRejection {
    fn into_response(self) -> axum::response::Response {
        match self {
            JsonDataError(inner) => <JsonDataError as IntoResponse>::into_response(inner),
            JsonSyntaxError(inner) => <JsonSyntaxError as IntoResponse>::into_response(inner),
            MissingJsonContentType(inner) => <MissingJsonContentType as IntoResponse>::into_response(inner),
            BytesRejection(inner) => <BytesRejection as IntoResponse>::into_response(inner),
        }
    }
}
```

Though the `struct` case could also be done via derive, with something like you showed. Not sure whether it's worth it if you need to write half the lines you'd write for the impl as attributes instead.

**Comment by davidpdrsn** [comment]
Yeah that would be

```rust
#[derive(IntoResponse)]
#[status = UNSUPPORTED_MEDIA_TYPE]
#[body = "Expected request with `Content-Type: text/sql`"]
pub struct MyRejectionReason;

// vs

pub struct MyRejectionReason;

impl IntoResponse for MyRejectionReason {
    fn into_response(self) -> Response {
        (
            StatusCode::UNSUPPORTED_MEDIA_TYPE,
            "Expected request with `Content-Type: text/sql`",
        ).into_response()
    }
}
```

**Comment by MRuecklCC** [comment]
IMHO declarative > imperative. I would 100% prefer the macro way :)
Is there interest in this? I may have a couple hours the next days to prepare a PR. But be warned, I am rather new to rust :P

**Comment by davidpdrsn** [comment]
I think the enum use case the jplatte shared in https://github.com/tokio-rs/axum/issues/1431#issuecomment-1262380059 is fine.

I don't think

```rust
#[derive(IntoResponse)]
#[status = UNSUPPORTED_MEDIA_TYPE]
#[body = "Expected request with `Content-Type: text/sql`"]
pub struct MyRejectionReason;
```

offers much over writing your own `IntoResponse` version.

**Comment by jplatte** [comment]
It is not possible for the proc-macro to read the value of the constant, but syntax-wise that should be no problem.
